### PR TITLE
Explore: Sub-tables support for Table component

### LINKED ISSUE #54997: Tempo - Collapsible trace rows in TraceQL table
Sync with the Grafana BI team (who own the [Table component](https://github.com/grafana/grafana/blob/main/packages/grafana-ui/src/components/Table/Table.tsx)) to add a way to configure [rows to be collapsible](https://react-table-v7.tanstack.com/docs/examples/expanding) or allow [expandable sub-components](https://react-table-v7.tanstack.com/docs/examples/sub-components).

Next, update the TraceQL response transformer to generate a table with a small set of spans inside the collapsable trace row. Regularly sync with the Tempo team and reach out to the UX team to find the best possible solution.

### PR DESCRIPTION
**Pull request by adrapereira**
**What is this feature?**

This PR adds sub-table support to the existing Table component, with the goal of displaying more information when the user opts to expand a row. 

This is achieved using the new `subData: DataFrame[]` prop of the Table component, where each Dataframe in the array links to a row of the "main" frame (under the `data` prop). The link is possible using the new custom meta property `parentRowIndex: number` that should exist in each frame of the `subData` array. 

```
const frame = new DataFrame({
   ...
   meta: {
      custom: {
         parentRowIndex: 12,
      },
   }
});
```

It's also possible to hide the header of sub-tables by setting the custom meta property `noHeader: boolean`

```
const frame = new DataFrame({
   ...
   meta: {
      custom: {
         noHeader: true,
      },
   }
});
```

This implementation only supports one level of sub-tables but it should be straightforward to add more levels, perhaps using a `depth: number` property and move to a list of parents instead of a single index. 

https://user-images.githubusercontent.com/4313971/203339553-56d90ec7-69b2-41e6-93e3-b0adcbd659e6.mov

Fixes https://github.com/grafana/grafana/issues/54997

### CODE REVIEW

File: public/app/plugins/panel/table/TablePanel.tsx [hunk 1]
```diff
@@ -123,8 +124,9 @@ export class TablePanel extends Component<Props> {
     const { data, height, width, options, fieldConfig, id } = this.props;
 
     const frames = data.series;
-    const count = frames?.length;
-    const hasFields = frames[0]?.fields.length;
+    const mainFrames = frames.filter((f) => f.meta?.custom?.parentRowIndex === undefined);
```
[the remark this answers is not in the document]
Comment (aocenas) [reply public/app/plugins/panel/table/TablePanel.tsx:127]: Yeah this is only thing that I am not sure about. So for example just to simulate multiple table dataframes you can do something like this:
Screenshot from 2022-11-22 18-27-19

In a dashboard just have 2 queries and then there is this frame selector in the table panel.

In Explore it seems we never actually showed 2 table data frames but we probably should somehow. In both cases though it seems we need to match the correct main and subframes together.

For that I think we can just use the `refId` https://github.com/grafana/grafana/blob/74c809f54459d235f4d3e9b1c117f291cbf16339/packages/grafana-data/src/types/data.ts#L124. That should be different for each query.

Not sure if there is an usecase for single query returning multiple tables with subrows, cannot think of anything right now, so the `refId` should be enough.

**Review by aocenas — APPROVED** [review]

**Review by codeincarnate — APPROVED** [review]
Checked through the changes and the functionality and it looks great! Don't have too much to add beyond what Andrej mentioned so looks good to me!

**Comment by torkelo** [comment]
@aocenas @codeincarnate 

wish I was pinged for review for this PR, feels like a column that is a nested DataFrame type would be a cleaner solution. (IE a field that is a nested data frame). Was this considered? 

Also the "Expand" action would be good to be able to do async, (And load the sub data async). is there a callback that is called when expanded to do that?

**Comment by aocenas** [comment]
> wish I was pinged for review for this PR, feels like a column that is a nested DataFrame type would be a cleaner solution. (IE a field that is a nested data frame). Was this considered?

Don't think so. Yeah I agree that sounds good but not sure if that has any implications on the DataFrame typings, mainly on the backend side whether we can easily recursively encode it into whatever arrow does on the backend.

> Also the "Expand" action would be good to be able to do async, (And load the sub data async). is there a callback that is called when expanded to do that?

I assume this could be an additional API? Personally I mostly think about how to drive these things from the perspective of datasource, which usually means dataFrame that can be passed from backend to frontend with minimal processing on the front end. From that perspective such thing may be interesting for table generally, I don't think it would be usable from datasource directly which was the main concern for us here.

**Comment by torkelo** [comment]
@aocenas I think the backend can supported nested dataframes , nothing preventing a field value to be a DataFrame (I think) 

Would require some type changes. But that is really needed to support sparklines cells (graphs) in a good way. 
I tried to explore that in this PR using simple Array field type but to do it properly need two fields (time + value), and a nested data frame wou
https://github.com/grafana/grafana/pull/41994 

The Expand callback is tricky it would be more for custom applications  / Scenes to use. 

Maybe there would be a way to do a expand callback via field custom option? So it would call into the frontend data source, that could then mutate the field values (fill them with data) :) Just brainstorming how we could do table drilldown where the drilldown data would be lazy :)

**Comment by torkelo** [comment]
@aocenas is there an Explore feature / data source that leverage this? How are the sub data queried for?

File: public/app/plugins/panel/table/TablePanel.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (torkelo) [reply public/app/plugins/panel/table/TablePanel.tsx:127]: @adrapereira can you describe how this works??

[the remark this answers is not in the document]
Comment (torkelo) [reply public/app/plugins/panel/table/TablePanel.tsx:127]: how is one to use this feature from dashboards? how do I set parentRowIndex?

**Comment by torkelo** [comment]
@adrapereira @aocenas is there a design doc for this? I am sorry for being a pain in...  but I am struggling a bit with this change. 

This hits at the problem of drill-downs that is so fundamental (and BI squad has been doing POCs / design docs etc around).  We have to solve and get this right but this feels a bit too ad hoc:
* Simple nested data but done in a very complex way with two separate data frame arrays linked by custom indices (instead of just making the data nested) 
* Requiring special data sources implementations to leverage (I guess this would be true for nested data frames as well). But more easier to make nested dataframes a generic concept that can be supported by other visualizations and tooling / transforms.   
* No way build on this to create async drilldowns (understandable as this tricky but also the really interesting challenge)

**Comment by torkelo** [comment]
All visualizations that does not check custom parentRowIndex will be very confused by this, and try to visualize the drilldown/detail data.

File: public/app/plugins/panel/table/TablePanel.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (torkelo) [reply public/app/plugins/panel/table/TablePanel.tsx:127]: @adrapereira how will this work for when tempo is used in Dashboards? for other visualizations than table?

This way to represent nested data feels very messy, needing special new filtering to split the data, work needed in every visualization to filter our data with parentRowIndex. 

Really wish this was done via design doc and shared a bit before, been thinking and POCing drilldown scenarios and how we can make Grafana great at exploring nested data sets for years, working with @dprokop in his efforts on drilldown queries etc. And this came a bit from nowhere (see comments further down in in PR). 

Table is just a core and important component, really want us to get nested data done right. Maybe it's this solution but would have loved if could have talked about it and explored other options.

**Comment by aocenas** [comment]
@torkelo the main driver for this was this https://github.com/grafana/grafana/pull/59313. TraceQL search that returns not only the matched trace but also the matched spans in that trace which we wanted to show somehow as that may be more interesting than the trace itself.

> How are the sub data queried for?

Basically, we get that already in a single response and we just pack it into multiple data frames. At this point this is on frontend but we should move that to backend.

>  is there a design doc for this?

There is short one (but not sure at this moment if it's up to date) https://docs.google.com/document/d/1Ae7Hir_lR8HxQIvOlwF0ZIKdnyr3NemSosY7JruWLJg/edit#heading=h.5sybau7waq2q and also we had few calls with the BI team to discuss the changes (earlier notes https://docs.google.com/document/d/1G-50bEdCslpdwFcB-RmrmZqTf44B8UOFGkJAl6RA3iI/edit#heading=h.bts1plh63p0z).

> Simple nested data but done in a very complex way with two separate data frame arrays linked by custom indices (instead of just making the data nested)

If nested dataFrame would be used to show a graph inside a cell then I would say this clashes with the subtables. Not sure how it would work with multiple subframe cells which does not make sense for this usecase. Seems like rendering a subframe inside the cell is a different abstraction (and different API) than rendering another set of rows that should be collapsed behind the parent row. We actually considered having a single data frame with some special config to mark what is shown in a suitable and what in main table but that ended up with much more complex table code while not providing a much better API.

> Requiring special data sources implementations to leverage (I guess this would be true for nested data frames as well). But more easier to make nested dataframes a generic concept that can be supported by other visualizations and tooling / transforms.

As you say don't seem like either solution would have any support right away, both need special DS code. From that stand point separate data frame with special attribute seems simpler IMHO (could be a separate query, just need something to link, could be a special field instead of config that would mark that second query data as child data).

> No way build on this to create async drilldowns

Not sure how this would work either way. I always thought dataFrame should be data only to be able to serialize it and send from backend. Adding stuff that is front end only and not serializable seems like going the opposite way of that.

**Comment by torkelo** [comment]
@aocenas 

> If nested dataFrame would be used to show a graph inside a cell then I would say this clashes with the subtables. Not sure how it would work with multiple subframe cells which does not make sense for this usecase

was thinking that the field would have a custom field config to mark to use it for a "nested table". It would have a DataFrame field type that would mean it was not shown as normal column anyway. So feels like a perfect fit for this feature, and how I always imaged doing nested data.

**Comment by torkelo** [comment]
another option could have been "nestedFrames" on the DataFrame itself. Just some way to pass nested data without requiring filtering in every visualization and new subData React props for every visualization

**Comment by aocenas** [comment]
> was thinking that the field would have a custom field config to mark...

I agree it's sort of neat but at the same I don't personally see a huge benefit. Mainly I would rather go in the direction of having API where you can just create 2 queries in a panel and with some simple naming conventions or single transform you can say it's a main frame and a subframe data. Not sure if dataFrame column type is closer to that.

> Just struggle with seeing how this solution is going to work in dashboards, as a generic solution without messing up all visualizations that would need to filter and segregate data based on meta fields and all who don't will be messed up. Where as a new field type that is not a string or number or time would be automatically ignored

I think I see what you mean. I could see how nested frames could be used to show some "detail" data in a node graph or a way to show exemplars and other things when clicking on a point in time series. I guess even table could be more flexible, and show multiple subtables when row was expanded, or even a full time series panel alongside that to build more rich subview :thinking:.

So I don't think it's a problem at this moment to change this. Shouldn't be breaking for Tempo. We did not advertise this API and it's marked as alpha. But I guess it's more of a BI squad call cc @grafana/grafana-bi-squad.

**Comment by torkelo** [comment]
@aocenas 

> I agree it's sort of neat but at the same I don't personally see a huge benefit. Mainly I would rather go in the direction of having API where you can just create 2 queries in a panel and with some simple naming conventions or single transform you can say it's a main frame and a subframe data. Not sure if dataFrame column type is closer to that.

Of curse it would still be two queries, and a transform. having nested data be passed in a NESTED WAY is all I am asking for (to panels/visualizations), not as flat DataFrames or via separate react props. As this requires every visualization to know about these meta fields to now visualize the nested data .

**Comment by torkelo** [comment]
nested data frames would also support multiple levels of nesting out of the box

**Comment by torkelo** [comment]
@aocenas @adrapereira 

Let's move this discussion to a meeting. I am really excited and happy that we are making some progress on nested data drilldown, having a expand action in the table to show nested data has been on my wishlist for years. Just want to make sure we design the data representation in a way that makes this work in other parts of Grafana (dashboards), and in a composable and nestable way (ie multiple levels), and in a way that adds minimal complexity and compatibility requirements to visualizations that don't support showing nested data.

**Comment by codeincarnate** [comment]
Thanks for all the very thoughtful feedback @torkelo! We had talked about async possibilities with this and I don't think the solution in question precludes that though I may be wrong there. I definitely hear you regarding simply nesting the data, although that seems like it could get a bit convoluted depending on how we end up nesting that data.

I'm definitely in favor of meeting about this so that we can figure out more of the details 😄

**Comment by torkelo** [comment]
@adrapereira Did you ever try using the built-in expanded feature in react-table ?

**Comment by torkelo** [comment]
Was easy to refactor this to use react-table internal state & our existing table reducer, simplified this quite a bit
https://github.com/grafana/grafana/pull/60791

**Pull request merged**